chain-signatures/node/src/indexer_eth/mod.rs

```diff
@@ -0,0 +1,1514 @@
+        else {
+            tracing::error!(
+                "Failed to parse contract address: {}",
+                eth_config_clone.contract_address
+            );
+            return;
+        };
+        let total_timeout = Duration::from_secs(self.eth.total_timeout);
+
+        let (blocks_failed_send, blocks_failed_recv) = failed_blocks_channel();
+
+        let (requests_indexed_send, requests_indexed_recv) = indexed_channel();
+
+        let (finalized_block_send, finalized_block_recv) = finalized_block_channel();
+
+        let (blocks_to_process_send, mut blocks_to_process_recv) = blocks_to_process_channel();
+
+        let client_clone = Arc::clone(&client);
+        let finalized_block_send_clone = finalized_block_send.clone();
+        let refresh_interval = self.eth.refresh_finalized_interval;
+        tokio::spawn(async move {
+            tracing::info!("Spawned task to refresh the latest finalized block");
+            Self::refresh_finalized_block(
+                client_clone,
+                finalized_block_send_clone,
+                refresh_interval,
+            )
+            .await;
+        });
+
+        let node_near_account_id_clone = self.node_near_account_id.clone();
+        let backlog_clone = self.backlog.clone();
+        let client_clone = Arc::clone(&client);
+        let optimistic_requests = self.eth.optimistic_requests;
+        let sign_tx_clone = sign_tx.clone();
+        tokio::spawn(async move {
+            Self::send_requests_when_final(
+                client_clone,
+                requests_indexed_recv,
+                finalized_block_recv,
+                sign_tx_clone,
+                app_data_storage.clone(),
+                node_near_account_id_clone,
+                optimistic_requests,
+                backlog_clone,
+            )
+            .await;
+        });
+
+        let blocks_failed_send_clone = blocks_failed_send.clone();
+        let node_near_account_id_clone2 = self.node_near_account_id.clone();
+        let requests_indexed_send_clone = requests_indexed_send.clone();
+        let backlog_clone2 = self.backlog.clone();
+        let sign_tx_clone = sign_tx.clone();
+        let client_clone = Arc::clone(&client);
+        tokio::spawn(async move {
+            Self::retry_failed_blocks(
+                client_clone,
+                blocks_failed_recv,
+                blocks_failed_send_clone,
+                contract_address,
+                node_near_account_id_clone2,
+                requests_indexed_send_clone,
+                total_timeout,
+                backlog_clone2,
+                sign_tx_clone,
+            )
+            .await;
+        });
+
+        let blocks_to_process_send_clone = blocks_to_process_send.clone();
+        let client_clone = Arc::clone(&client);
+        if let Some(last_processed_block) = last_processed_block {
+            match Self::catchup_end_block_number(client_clone).await {
```

Comment:
nit: since this is in an if statement, only clone inside of it.

```suggestion
        if let Some(last_processed_block) = last_processed_block {
            match Self::catchup_end_block_number(Arc::clone(&client)).await {
```